### PR TITLE
MULE-13979: Find a workaround for not having to detect policy logger …

### PR DESCRIPTION
**Pull request by pablokraan**
…context in runtime

### CODE REVIEW

File: modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java [hunk 1]
```diff
@@ -94,23 +100,43 @@ public void removeContext(LoggerContext context) {
    * @return the {@link ClassLoader} owner of the logger context
    */
   static ClassLoader resolveLoggerContextClassLoader(ClassLoader classLoader) {
-    ClassLoader loggerClassLoader = classLoader == null ? Thread.currentThread().getContextClassLoader() : classLoader;
+    WeakReference<ClassLoader> loggerClassLoaderReference = classLoaderMap.get(classLoader);
+    ClassLoader loggerClassLoader;
+
+    if (loggerClassLoaderReference == null) {
+      loggerClassLoader = getLoggerClassLoader(classLoader);
```
Comment (elrodro83) [inline modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java]: even though the map is synchronized, this code may be executed concurrently and the loggerClassLoader calculated extra times needlessly.

File: modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java [hunk 2]
```diff
@@ -56,6 +60,8 @@
   private final MuleLoggerContextFactory loggerContextFactory = new MuleLoggerContextFactory();
 
   private LoggerContextCache cache = new LoggerContextCache(this, getClass().getClassLoader());
+  private static final Map<ClassLoader, WeakReference<ClassLoader>> classLoaderMap =
```
Comment (elrodro83) [inline modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java]: rename to `loggerClassLoaders` or `loggerClassLoadersMap`

File: modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java (hunk 1 above)
Comment (pablokraan) [reply modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java]: sure, but that is a minor performance overhead, I can't think of any solution that will avoid that duplicated calculation without requiring a locking, which will be worst that this solution. Do you have an alternative solution?

Comment (pablokraan) [reply modules/launcher/src/main/java/org/mule/runtime/module/launcher/log4j2/ArtifactAwareContextSelector.java]: disregard my previous comment...

**Pull request merged**